File: rust/mapbox/vector_layer.rs
use crate::{
    base::BaseVectorLayer,
    mapbox::{write_feature, MapboxVectorFeature},
    VectorFeatureMethods, VectorLayerMethods,
};
use alloc::{
    collections::{btree_map::Entry, BTreeMap},
    rc::Rc,
    string::String,
    vec::Vec,
};
use core::cell::RefCell;
use pbf::{ProtoRead, Protobuf};
use s2json::PrimitiveValue;

/// Mapbox specification for a Layer
#[derive(Debug)]
pub struct MapboxVectorLayer {
    /// the version of the vector tile layer.
    pub version: u16,
    /// the name of the layer
    pub name: String,
    /// the extent of the vector layer
    pub extent: usize,
    /// the features in the layer
    pub features: BTreeMap<usize, MapboxVectorFeature>,
    /// track the positions of the features
    pub feature_positions: Vec<usize>,
    /// whether or not the layer is an s2 layer. This is an extension to the Mapbox spec and not used
    /// in production by most tools
    is_s2: bool,
    /// a reference to the pbf
    pbf: Rc<RefCell<Protobuf>>,
    /// key store used by features
    keys: Rc<RefCell<Vec<String>>>,
    /// value store used by features
    values: Rc<RefCell<Vec<PrimitiveValue>>>,
}
impl MapboxVectorLayer {
    /// Create a new MapboxVectorLayer
    pub fn new(pbf: Rc<RefCell<Protobuf>>, is_s2: bool) -> MapboxVectorLayer {
        MapboxVectorLayer {
            version: 5,
            name: String::new(),
            extent: 4_096,
            is_s2,
            pbf: pbf.clone(),
            keys: Rc::new(RefCell::new(Vec::new())),
            values: Rc::new(RefCell::new(Vec::new())),
            features: BTreeMap::new(),
            feature_positions: Vec::new(),
        }
    }
}
impl VectorLayerMethods for MapboxVectorLayer {
    fn version(&self) -> u16 {
        self.version
    }

    fn name(&self) -> String {
        self.name.clone()
    }

    fn extent(&self) -> usize {
        self.extent
    }

    /// the number of features in the layer
    fn len(&self) -> usize {
        self.feature_positions.len()
    }

    /// Check if the layer is empty
    fn is_empty(&self) -> bool {
        self.feature_positions.is_empty()
    }

    fn feature(&mut self, i: usize) -> Option<&mut dyn VectorFeatureMethods> {
        // First check if self.features already has the feature
        if let Entry::Vacant(e) = self.features.entry(i) {
            // Read the feature
            let mut feature = MapboxVectorFeature::new(
                self.pbf.clone(),
                self.is_s2,
                self.extent,
                self.version,
                self.keys.clone(),
                self.values.clone(),
            );
            let mut pbf = self.pbf.borrow_mut();
            pbf.set_pos(self.feature_positions[i]);
            pbf.read_message(&mut feature);
            e.insert(feature);

            // Now safely retrieve the inserted feature
            Some(self.features.get_mut(&i).unwrap() as &mut dyn VectorFeatureMethods)
        } else {
            // Safe to unwrap since we just checked the key exists
            Some(self.features.get_mut(&i).unwrap() as &mut dyn VectorFeatureMethods)
        }
    }
}
impl ProtoRead for MapboxVectorLayer {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            15 => self.version = pb.read_varint::<u16>(),
            1 => self.name = pb.read_string(),
            2 => self.feature_positions.push(pb.get_pos()),
            3 => {
                self.keys.borrow_mut().push(pb.read_string());
            }
            4 => {
                let mut value = PrimitiveValue::Null;
                pb.read_message(&mut value);
                self.values.borrow_mut().push(value);
            }
            5 => self.extent = pb.read_varint::<usize>(),
            #[tarpaulin::skip]
            _ => panic!("Unknown layer type"),
        }
    }
}

/// Write a layer to a protobuffer using the S2 Specification
pub fn write_layer(layer: &BaseVectorLayer, mapbox_support: bool) -> Vec<u8> {
    let mut pbf = Protobuf::new();
    let mut keys: BTreeMap<String, usize> = BTreeMap::new();
    let mut values: BTreeMap<PrimitiveValue, usize> = BTreeMap::new();

    pbf.write_varint_field(15, if mapbox_support { 1 } else { 5 });
    pbf.write_string_field(1, &layer.name);
    for feature in layer.features.iter() {
        pbf.write_bytes_field(2, &write_feature(feature, &mut keys, &mut values, mapbox_support));
    }
    let mut keys: Vec<(String, usize)> = keys.into_iter().collect();
    keys.sort_by(|a, b| a.1.cmp(&b.1));
    // keys and values
    for (key, _) in keys.iter() {
        pbf.write_string_field(3, key);
    }
    let mut values: Vec<(PrimitiveValue, usize)> = values.into_iter().collect();
    values.sort_by(|a, b| a.1.cmp(&b.1));
    for (value, _) in values.iter() {
        pbf.write_message(4, value);
    }
    pbf.write_varint_field(5, layer.extent as usize);

    pbf.take()
}

File: src/mapbox/vectorLayer.ts
import VectorFeature from './vectorFeature';

import type { Extents } from '../open';
import type { Pbf as Protobuf } from '../pbf';
import type { Value } from '../vectorTile.spec';

/**
 * A MapboxVectorLayer is a storage structure for the vector tile.
 * It may contain either the old Mapbox layers or the new S2 layers.
 * Parses extent, keys, values, and features. Features will utilize the extent, keys, and values.
 */
export default class MapboxVectorLayer {
  version = 5;
  name = 'default';
  extent: Extents = 4_096;
  length = 0;
  isS2: boolean;
  #pbf: Protobuf;
  #keys: string[] = [];
  #values: Value[] = [];
  #featuresPos: number[] = [];
  #features = new Map<number, VectorFeature>();
  /**
   * @param pbf - The Protobuf object to read from
   * @param end - The end position of the message in the buffer
   * @param isS2 - Whether the layer is an S2 layer or Mapbox layer
   */
  constructor(pbf: Protobuf, end: number, isS2 = false) {
    this.#pbf = pbf;
    this.isS2 = isS2;
    pbf.readFields(this.#readLayer, this, end);
    this.length = this.#featuresPos.length;
  }

  /**
   * @param tag - The tag of the message
   * @param layer - The layer to mutate
   * @param pbf - The Protobuf object to read from
   */
  #readLayer(tag: number, layer: MapboxVectorLayer, pbf: Protobuf): void {
    if (tag === 15) layer.version = pbf.readVarint();
    else if (tag === 1) layer.name = pbf.readString();
    else if (tag === 2) layer.#featuresPos.push(pbf.pos);
    else if (tag === 3) layer.#keys.push(pbf.readString());
    else if (tag === 4) layer.#values.push(layer.#readValueMessage(pbf));
    else if (tag === 5) layer.extent = pbf.readVarint() as Extents;
  }

  /**
   * @param i - The index of the feature
   * @returns - A feature at the given index
   */
  feature(i: number): VectorFeature {
    if (i < 0 || i >= this.#featuresPos.length) throw new Error('feature index out of bounds');
    let vtf = this.#features.get(i);
    if (vtf !== undefined) return vtf;

    this.#pbf.pos = this.#featuresPos[i];
    const end = this.#pbf.readVarint() + this.#pbf.pos;
    vtf = new VectorFeature(
      this.#pbf,
      end,
      this.isS2,
      this.extent,
      this.version,
      this.#keys,
      this.#values,
    );

    this.#features.set(i, vtf);

    return vtf;
  }

  /**
   * @param pbf - The Protobuffer object to read from
   * @returns - A parsed Value
   */
  #readValueMessage(pbf: Protobuf): Value {
    let value: Value = null;

    const end = pbf.readVarint() + pbf.pos;

    while (pbf.pos < end) {
      const tag = pbf.readVarint() >> 3;

      if (tag === 1) value = pbf.readString();
      else if (tag === 2) value = pbf.readFloat();
      else if (tag === 3) value = pbf.readDouble();
      else if (tag === 4) value = pbf.readVarint64();
      else if (tag === 5) value = pbf.readVarint();
      else if (tag === 6) value = pbf.readSVarint();
      else if (tag === 7) value = pbf.readBoolean();
      else value = null;
    }

    return value;
  }
}
